src/Symfony/Component/Security/OAuth2Client/Provider/GenericProvider.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\OAuth2Client\Provider;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Security\OAuth2Client\Exception\InvalidRequestException;
+use Symfony\Component\Security\OAuth2Client\Exception\InvalidUrlException;
+use Symfony\Component\Security\OAuth2Client\Loader\ClientProfileLoader;
+use Symfony\Component\Security\OAuth2Client\Token\RefreshToken;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @author Guillaume Loulier <contact@guillaumeloulier.fr>
+ */
+abstract class GenericProvider implements ProviderInterface
+{
+    private const DEFAULT_OPTIONS = [
```

Comment:
I think that some of those keys are not always required including `redirect_uri` according the use case (client credentials flow for example).

Unless I miss something, it may concern the option `userDetails_url` too as a user is not always involved.